torch/csrc/jit/codegen/cuda/tensor_view.cpp

```diff
@@ -891,6 +893,14 @@ void TensorView::clearReductionIterDomains() {
   setDomain(IrBuilder::create<TensorDomain>(container(), new_root, new_contig));
 }
 
+void TensorView::doubleBuffer() {
+  auto def = dynamic_cast<UnaryOp*>(definition());
+  TORCH_CHECK(
+      def != nullptr && def->getUnaryOpType() == UnaryOpType::Set,
+      "Double buffering is only supported when a TensorView is defined in a fusuion with the Set unary op.");
```

Comment:
I saw the checks in lower_shift on memory types. I think it would be good to have them here as well so we can error during scheduling if anything is wrong.

Reply:
I didn't check the memory type here so that `TensorView::setMemoryType()` could be done after `TensorView::doubleBuffer()`. I thought that's more consistent with, say, parallelization where shared memory needs to be used, but we don't check the memory type when setting parallel types. The only condition that won't change at this point is how the tensor is defined, so that's why it's the only one checked here.

I don't have a strong opinion here. Let me know if the memory type should be checked here.